### PR TITLE
Fix missing element types and shapes when autograd.Function has multiple tensor outputs

### PR DESCRIPTION
**Pull request by wschin**
When generating IR for autograd.Function, if the function has multiple outputs, a TupleUnpack may be inserted after the original function node, and Pytorch only assigns proper information (tensor element type and shape) to the TupleUnpack and forgets the original function node. In contrast, if autograd.Function only produces one output, the original function node may have tensor 
element type and shape in its output schema.

Before this PR:
- (simplified) IR for autograd.Function with one output: input (tensor, dtype=float32, shape=[2, 3]) -> PythonOp -> output (tensor, dtype=float32, shape=[4, 5])
- (simplified) IR for autograd.Function with one output: input (tensor, dtype=float32, shape=[2, 3]) -> PythonOp -> output_0 **(tensor)**, output_1 **(tensor)** -> TupleUnpack output_2 (tensor, dtype=float32, shape=[4, 5]), output_3 (tensor, dtype=float32, shape=[6, 7])

After this PR:
- (simplified) IR for autograd.Function with one output: input (tensor, dtype=float32, shape=[2, 3]) -> PythonOp -> output (tensor, dtype=float32, shape=[4, 5])
- (simplified) IR for autograd.Function with one output: input (tensor, dtype=float32, shape=[2, 3]) -> PythonOp ->output_0 **(tensor, dtype=float32, shape=[4, 5])**, output_1 **(tensor, dtype=float32, shape=[6, 7])** -> TupleUnpack output_2 (tensor, dtype=float32, shape=[4, 5]), output_3 (tensor, dtype=float32, shape=[6, 7])

### CODE REVIEW

**Comment by albanD** [comment]
cc @eellison do you know who could review this?

**Comment by eellison** [comment]
cc @gmagogsfm

**Comment by BowenBao** [comment]
cc @gmagogsfm @SplitInfinity friendly ping for review 

Usually for ONNX related updates we review and merge into onnx_ms_1 branch, but since this one interacts with autograd function and how it's traced, I think it's best for you to take a look.

File: torch/csrc/autograd/python_function.cpp
Comment (BowenBao) [inline torch/csrc/autograd/python_function.cpp:520]:
```suggestion
  // If TupleUnpack operator is created, we copy its output type back
```

**Comment by BowenBao** [comment]
@gmagogsfm ping ...

File: test/jit/test_tracer.py
Comment (wschin) [inline test/jit/test_tracer.py:1412]:
```suggestion
        # handling is correct in TorchScript.
```

File: test/jit/test_tracer.py
Comment (gmagogsfm) [inline test/jit/test_tracer.py:1409]: Does this test fail without shape information on original call note?

Comment (wschin) [reply test/jit/test_tracer.py:1409]: It doesn't but it still checks the newly added part. If new change is wrong (I tried randomly assigned some shapes), this test may fail.

Comment (gmagogsfm) [reply test/jit/test_tracer.py:1409]: In that case, could you use `FileCheck()` to ensure correct tensor information is in the IR? That would make the test fail in case of tensor properties being propagated differently

Comment (wschin) [reply test/jit/test_tracer.py:1409]: I have added new logics to directly examine my change:
```
        for n in traced.graph.nodes():
            if n.kind() != 'prim::PythonOp':
                continue
            for o in n.outputs():
                if o.type().kind() != 'TupleType':
                    continue
                for e in o.type().elements():
                    # In this test, tensor elements in a tuple output 
                    # can't have unknown shapes.
                    assert e.dim()
```
This `assert` throws without my change. Basically, I check the IR before dumping it to file, so we don't need `FileCheck()` here.

Comment (pengwa) [reply test/jit/test_tracer.py:1409]: Hi @gmagogsfm Could you please help review this change? thanks a lot.

File: test/jit/test_tracer.py
Comment (gmagogsfm) [inline test/jit/test_tracer.py:1443]: Nit:
Please use `self.assert***` methods instead of raw `assert`

File: test/jit/test_tracer.py
Comment (gmagogsfm) [reply test/jit/test_tracer.py:1409]: `FileCheck()` doesn't require dumping IR to a file actually, you can pass a string to it, it is generally nicer to write and read than logics that directly check for a pattern.

Here is an exmaple:
https://github.com/pytorch/pytorch/blob/22e3cc21e535e29deea091adf93e0408139b6aa3/test/test_jit.py#L496

**Review by gmagogsfm — APPROVED** [review]
I think this is functionally sound, but there are a few nits on how the tests are written.

Sorry for the delay, message was buried in floods of GH emails and this PR dropped off from my GH review dashboard since last round of comment.

File: test/jit/test_tracer.py
Comment (wschin) [reply test/jit/test_tracer.py:1443]: I am trying to replace those checks with `FileCheck`, so these `assert`s will be removed. Thanks.

File: test/jit/test_tracer.py
Comment (wschin) [reply test/jit/test_tracer.py:1409]: Done using `FileCheck`. Thanks.

**Comment by wschin** [comment]
@gmagogsfm, Thanks a lot for reviewing again. All comments are addressed now. Once the build passes, could you please merge this PR? Please let me know if concerns.

**Review by gmagogsfm — APPROVED** [review]

**Pull request closed without merging**
